### DISCUSSION TITLE
Problema de Autenticação com Blazor WASM e API Externa

### DISCUSSION THREAD

**Discussion by LucasBLs** [balta-io/forum#636] [Tira Dúvidas] [answered]
CURSO FULL STACK

**Contexto:**
Estou testando a aplicação front-end em Blazor WebAssembly que se conecta a uma API externa hospedada em um servidor. A aplicação continua utilizando autenticação baseada em cookies, conforme o curso.

**Descrição do Problema:**
Quando tento fazer login e acessar endpoints protegidos da API a partir do Blazor WASM, ocorre um refresh na tela e recebo um erro 401 Unauthorized no endpoint `/v1/identity/manage/info` no console, não aparece nenhum snackbar de erro. No entanto, quando executo a mesma aplicação via `dotnet run`, a autenticação funciona corretamente.

**Detalhes Técnicos:**

- **API:** .NET Core API hospedada em cloud.
- **Front-End:** Blazor WebAssembly (Hospedado no IIS do windows).
- **Autenticação:** Baseada em cookies.
- **CORS:** Configurado na API para permitir origens específicas (Não há erros de cors no console).
- **Erro:** 401 Unauthorized ao acessar endpoints.

**Exceção e Logs:**
```
GET http://backend-url.com/v1/identity/manage/info 401 (Unauthorized)
System.Net.Http.HttpRequestException: net_http_message_not_success_statuscode_reason, 401, Unauthorized
```

Observei que o retorno do login é 200, porém após entrar no   `  await AuthenticationStateProvider.GetAuthenticationStateAsync();`
ele cai nesse metodo retornando o 401:
```
   private async Task<User?> GetUser()
   {
       try
       {
           return await _httpClient.GetFromJsonAsync<User?>("v1/identity/manage/info");
       }
       catch (Exception ex)
       {
           Console.WriteLine(ex);
           return null;
       }
   }
```

No OnInitializedAsynco user não está autenticado, porém o mesmo usuário e senha eu tentei via postman e deu certo...

```
  protected override async Task OnInitializedAsync()
  {
      var authState = await AuthenticationStateProvider.GetAuthenticationStateAsync();
      var user = authState.User;

      if (user.Identity != null && user.Identity.IsAuthenticated)
          Navigation.NavigateTo("/");
  }
```

**Passos Tentados:**

1. Verifiquei se o token de autenticação e os cookies estão sendo enviados corretamente.
2. Garanti que a configuração de CORS na API permite as origens do front-end e back-end.
3. Testei a autenticação manualmente usando Postman, que funcionou conforme esperado.

**Questão:**
Alguém já enfrentou um problema similar ao conectar um front-end Blazor WASM a uma API externa utilizando autenticação por cookies? Existe alguma configuração específica ou ajuste que pode estar faltando?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @LucasBLs, como vai?

Primeiramente, parabéns por usar o template e detalhar a dúvida, você ajuda muito nosso trabalho 💜

Sobre seu problema, vamos lá:

A primeira possível causa é a porta de execução das aplicações. Quando rodamos em debug ele pode rodar em HTTP ou HTTPS, e sim, isto influencia no CORS.

As portas são definidas no arquivo `launchSettings.json` na pasta `Properties`, como aqui por exemplo:
https://github.com/balta-io/3054/blob/main/src/Dima.Web/Properties/launchSettings.json

### Como resolver
Procure todas as entradas para `"applicationUrl": "http://localhost:XXXXX"` no projeto `Dima.Web` e anote todas estas URLs.
No projeto `Dima.Api`, no arquivo `BuilderExtension.cs` na pasta `Common/Api`, adicione estas URLs ao método `AddCrossOrigin`:
https://github.com/balta-io/3054/blob/main/src/Dima.Api/Common/Api/BuilderExtension.cs#L57

O resultado final deve ficar parecido com isto:
```csharp
public static void AddCrossOrigin(this WebApplicationBuilder builder)
    {
        builder.Services.AddCors(
            options => options.AddPolicy(
                ApiConfiguration.CorsPolicyName,
                policy => policy
                    .WithOrigins([
                        Configuration.BackendUrl,
                        Configuration.FrontendUrl,
"NOVA URL",
"NOVA URL",
"NOVA URL",
...
                    ])
                    .AllowAnyMethod()
                    .AllowAnyHeader()
                    .AllowCredentials()
            ));
    }
```

Testa e me retorna!

**Comment by claudiavieira** [comment]
Olá, boa noite!
Estou exatamente com o mesmo problema do Lucas. Já fiz o procedimento acima, mas continua da mesma forma.
OBS: Se eu executar via prompt de comando, "dotnet run" nos projetos Dima.Api e Dima.Web, consigo efetuar o login, porém, eu gostaria de depurar com breakpoints o projeto.
@andrebaltieri sabe o que pode ser?
@LucasBLs voce conseguiu solucionar?

**Reply by andrebaltieri** [reply]
O problema é que as portas mudam dependendo do jeito que você roda...
Inicia o projeto API no modo debug pela IDE que você utilizar e veja a porta que ele roda.

Anota esta porta e usa ela nas configurações da API/APP que vai funcionar!
